### DISCUSSION TITLE
Using OPTIONAL in function definitions

### DISCUSSION THREAD

**Discussion by adriangb** [geldata/gel#2100] [Q&A] [answered]
I'm posting in discussion because I feel like I must be doing something wrong.

I tested that:
```
SELECT (<int64>{} ?? 1);
```
Gives:
```
{1}
```

So I tried to define a function:
```
function test_func(param: OPTIONAL int64) -> int64 { USING (SELECT (param ?? 1)) };
```

But when I call:

```
SELECT test_func();
```

I get:

```
error: could not find a function variant test_func
  ┌─ query:1:8
  │
1 │ SELECT test_func();
  │        ^^^^^^^^^^^^ error
```

**Comment by elprans** [comment, marked as answer]
`OPTIONAL` relates to the *type* of the arguments and signifies that it can be an empty set.  To make the argument as a whole optional you need to specify a default value:

`function test_func(param: OPTIONAL int64 = {}) -> int64 { USING (SELECT (param ?? 1)) };`

**Reply by 1st1** [reply]
I'm wondering if we could make the error message clearer though. Perhaps instead of

```
error: could not find a function variant test_func
```

it should be 

```
error: the "test_func" function expects one argument, none were provided
```

**Reply by elprans** [reply]
This is what https://github.com/edgedb/edgedb/pull/2083 is supposed to fix

**Comment by adriangb** [comment]
Excellent, thank you!
